#Document retrieval from wikipedia data
##Fire up pandas

In [1]:
import pandas
import sklearn
import numpy as np

##Load some text data - from wikipedia, pages on people

In [2]:
people = pandas.read_csv('people_wiki.csv')

Data contains: link to wikipedia article, name of person, text of article.

In [3]:
people.head()

URI                 name  \
0        <http://dbpedia.org/resource/Digby_Morrell>        Digby Morrell   
1       <http://dbpedia.org/resource/Alfred_J._Lewy>       Alfred J. Lewy   
2        <http://dbpedia.org/resource/Harpdog_Brown>        Harpdog Brown   
3  <http://dbpedia.org/resource/Franz_Rottensteiner>  Franz Rottensteiner   
4               <http://dbpedia.org/resource/G-Enka>               G-Enka   

                                                text  
0  digby morrell born 10 october 1979 is a former...  
1  alfred j lewy aka sandy lewy graduated from un...  
2  harpdog brown is a singer and harmonica player...  
3  franz rottensteiner born in waidmannsfeld lowe...  
4  henry krvits born 30 december 1974 in tallinn ...

In [4]:
len(people)

59071

##Explore the dataset and checkout the text it contains
###Exploring the entry for president Obama

In [5]:
obama = people[people['name'] == 'Barack Obama']

In [6]:
obama

URI          name  \
35817  <http://dbpedia.org/resource/Barack_Obama>  Barack Obama   

                                                    text  
35817  barack hussein obama ii brk husen bm born augu...

In [7]:
obama['text'].values

array([ 'barack hussein obama ii brk husen bm born august 4 1961 is the 44th and current president of the united states and the first african american to hold the office born in honolulu hawaii obama is a graduate of columbia university and harvard law school where he served as president of the harvard law review he was a community organizer in chicago before earning his law degree he worked as a civil rights attorney and taught constitutional law at the university of chicago law school from 1992 to 2004 he served three terms representing the 13th district in the illinois senate from 1997 to 2004 running unsuccessfully for the united states house of representatives in 2000in 2004 obama received national attention during his campaign to represent illinois in the united states senate with his victory in the march democratic party primary his keynote address at the democratic national convention in july and his election to the senate in november he began his presidential campaign in 2007 

###Exploring the entry for actor George Clooney

In [8]:
clooney = people[people['name'] == 'George Clooney']
clooney['text'].values

array([ 'george timothy clooney born may 6 1961 is an american actor writer producer director and activist he has received three golden globe awards for his work as an actor and two academy awards one for acting and the other for producingclooney made his acting debut on television in 1978 and later gained wide recognition in his role as dr doug ross on the longrunning medical drama er from 1994 to 1999 for which he received two emmy award nominations while working on er he began attracting a variety of leading roles in films including the superhero film batman robin 1997 and the crime comedy out of sight 1998 in which he first worked with a director who would become a longtime collaborator steven soderbergh in 1999 clooney took the lead role in three kings a wellreceived war satire set during the gulf warin 2001 clooneys fame widened with the release of his biggest commercial success the heist comedy oceans eleven the first of the film trilogy a remake of the 1960 film with frank sina

###Get the word counts for Obama article

In [9]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
features = vectorizer.fit_transform(obama['text'])

In [10]:
features

<1x270 sparse matrix of type '<type 'numpy.int64'>'
	with 270 stored elements in Compressed Sparse Row format>

There must be 270 words since we have 270 elements in our vector

In [11]:
vocab = vectorizer.get_feature_names()

In [12]:
print vocab

[u'13th', u'1961', u'1992', u'1996', u'1997', u'20', u'2000in', u'2004', u'2007', u'2008', u'2009', u'2010', u'2011', u'2012', u'2012obama', u'2013', u'44th', u'63', u'act', u'address', u'administration', u'affordable', u'afghanistan', u'african', u'after', u'against', u'american', u'americans', u'and', u'arms', u'as', u'ask', u'at', u'attention', u'attorney', u'august', u'barack', u'before', u'began', u'bin', u'bm', u'born', u'briefs', u'brk', u'budget', u'by', u'californias', u'called', u'campaign', u'care', u'chicago', u'civil', u'clinton', u'close', u'columbia', u'combat', u'community', u'constitutional', u'consumer', u'continued', u'control', u'convention', u'court', u'creation', u'cuba', u'current', u'death', u'debate', u'debt', u'defeated', u'defeating', u'defense', u'degree', u'delegates', u'democratic', u'district', u'doddfrank', u'domestic', u'dont', u'down', u'during', u'earning', u'economic', u'election', u'elementary', u'ended', u'ending', u'equality', u'federal', u'filed'

In [13]:
dist = np.sum(features.toarray(), axis=0)
dist

array([ 1,  1,  1,  1,  1,  2,  1,  3,  1,  1,  3,  2,  3,  1,  1,  1,  1,
        1,  8,  1,  1,  1,  2,  1,  4,  1,  3,  1, 21,  1,  6,  1,  2,  1,
        1,  1,  1,  1,  1,  1,  1,  2,  1,  1,  1,  1,  1,  1,  3,  1,  2,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  4,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  4,  1,  1,  2,  2,  1,  2,  1,  1,  3,  1,
        1,  1,  1,  1,  1,  3,  4,  2,  1,  3,  1,  1,  1,  1,  1,  1,  2,
        4,  1,  7,  1, 11,  1,  1,  1,  2,  1,  1,  1,  2, 30,  1,  1,  1,
        1,  1,  1,  3,  4,  2,  1,  3,  1,  1,  1,  1,  1,  1,  6,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  4,  1,  1,  1,  2,  1,  1,
        1,  1,  1,  2,  1,  1,  2,  9,  1, 18,  2,  1,  2,  1,  1,  1,  3,
        1,  1,  1,  1,  3,  1,  1,  1,  2,  4,  2,  1,  2,  1,  1,  1,  1,
        2,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  1,  1,
        2,  1,  2,  1,  3,  1,  1,  1,  1,  1,  1,  1,  1,  3,  1,  2,  3,
        2,  1,  3,  1,  1

## Sort the word counts for the Obama article
Turn the word counts into a pandas dataframe

In [14]:
obama_word_count_table = pandas.DataFrame({'token':vocab, 'count':dist}).sort('count')

In [15]:
obama_word_count_table.sort('count', ascending=False).head(15)

count       token
242     40         the
115     30          in
28      21         and
162     18          of
245     14          to
106     11         his
160      9       obama
18       8         act
104      7          he
133      6         law
256      6          us
30       6          as
260      5         was
74       4  democratic
123      4        iraq

Most common words include uninformative words like "the", "in", "and",...

##Compute TF-IDF for the corpus
To give more weight to informative words, we weigh them by their TF-IDF scores.

We'll vectorize the entire corpus and compute the TF-IDF in one step

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer
vect = TfidfVectorizer()

In [17]:
dtm = vect.fit_transform(people['text'])

In [18]:
dtm.shape

(59071, 548429)

In [19]:
features = vect.get_feature_names()

###Examine the TF-IDF for the Obama article
Remember that the Obama article is at index 35817

In [20]:
obama

URI          name  \
35817  <http://dbpedia.org/resource/Barack_Obama>  Barack Obama   

                                                    text  
35817  barack hussein obama ii brk husen bm born augu...

So we'll examine the document term matrix at index 35817 and pull out the TF-IDF by each word.

First we'll make a dictionary

In [21]:
word_scores = {}
for word in vocab:
    word_scores[word] = dtm[35817, features.index(word)]

And then convert it into a DataFrame

In [22]:
obama_word_scores = pandas.DataFrame(word_scores.items(), columns=['word', 'tfidf'])

In [23]:
obama_word_scores.sort('tfidf', ascending=False).head(15)

word     tfidf
241        obama  0.365018
62           the  0.279323
126          act  0.249089
218           in  0.209673
152         iraq  0.151809
204          and  0.146739
253          law  0.144687
132      control  0.131857
120           of  0.126205
55            us  0.122834
206      ordered  0.122380
127     military  0.119498
257   democratic  0.114575
255     response  0.110187
250  involvement  0.110187

For some reason common words are ranked fairly high, such as "*the*", "*in*", and "*of*" . I am not sure why this is the case, as I would expect the TF IDF to weight them much lower. If I had to take a guess, scikit-learn is probably expecting us to include the stop words when we vectorize the documents.

### Try again without stop words

We can try vectorizing everything again, but exclude any "stop words" (common words such as "the", "in" or "of")

In [24]:
vect = TfidfVectorizer(stop_words = 'english')
dtm = vect.fit_transform(people['text'])
features = vect.get_feature_names()

In [25]:
dtm.shape

(59071, 548116)

We now have 548,116 words instead of 548,429 since we don't care about the stop words

In [26]:
word_scores = {}
for word in vocab:
    if word in features: # vocab contains stop words that are no longer in our dtm
        word_scores[word] = dtm[35817, features.index(word)]

In [27]:
obama_word_scores = pandas.DataFrame(word_scores.items(), columns=['word', 'tfidf'])

In [28]:
obama_word_scores.sort('tfidf', ascending=False).head(15)

word     tfidf
157        obama  0.413495
109          act  0.282170
129         iraq  0.171970
213          law  0.163903
173      control  0.149369
175      ordered  0.138633
110     military  0.135368
216   democratic  0.129792
210  involvement  0.124821
215     response  0.124821
53        senate  0.104114
94          term  0.097433
115     campaign  0.095529
176      nominee  0.090386
67   afghanistan  0.090297

This matches pretty closely with the results from GraphLab Create

### An alternative approach
Although below we're only computing the TF IDF on the Obama article, not the entire corpus of all articles, so we won't use this in the rest of our analysis, but I included it for reference.

In [29]:
tfidf_obama_only = TfidfVectorizer(stop_words = 'english')
vectorized_obama_only = tfidf_obama_only.fit_transform(obama['text'])
tfidf_obama = zip(tfidf_obama_only.get_feature_names(), np.asarray(vectorized_obama_only.sum(axis=0)).ravel())
tfidf_obama.sort(key=lambda x: x[1], reverse=True)
tfidf_obama[0:15]

[(u'obama', 0.34188172937891387),
 (u'act', 0.30389487055903452),
 (u'law', 0.22792115291927589),
 (u'control', 0.15194743527951726),
 (u'democratic', 0.15194743527951726),
 (u'iraq', 0.15194743527951726),
 (u'military', 0.15194743527951726),
 (u'president', 0.15194743527951726),
 (u'2004', 0.11396057645963795),
 (u'2009', 0.11396057645963795),
 (u'2011', 0.11396057645963795),
 (u'american', 0.11396057645963795),
 (u'campaign', 0.11396057645963795),
 (u'election', 0.11396057645963795),
 (u'involvement', 0.11396057645963795)]

#Manually compute similarity and distances between a few people

Let's manually compare the similarity and distances between the articles for a few famous people.  

In [30]:
from sklearn.metrics.pairwise import cosine_similarity
clinton = people[people['name'] == 'Bill Clinton']
beckham = people[people['name'] == 'David Beckham']

##Is Obama closer to Clinton than to Beckham?

Note that we're computing the cosine **similarity** (which ranges from 0 to 1, where 1 is most similar) not the cosine **distance** which we can use as well.

cosine_distance = (1-cosine_similarity) 

We should find that the article about president Obama is closer to the one about former president Clinton than that of footballer David Beckham.

In [31]:
cosine_similarity(dtm[obama.index], dtm[clinton.index])

array([[ 0.18896718]])

In [32]:
cosine_similarity(dtm[obama.index], dtm[beckham.index])

array([[ 0.02556581]])

Indeed Obama and Clinton are more similar (0.188) than Obama and Beckham (0.02)

But we can also compute the cosine **distance** as we did in GraphLab Create

In [33]:
from sklearn.metrics.pairwise import pairwise_distances
pairwise_distances(dtm[obama.index], dtm[clinton.index], metric='cosine')

array([[ 0.81103282]])

In [34]:
pairwise_distances(dtm[obama.index], dtm[beckham.index], metric='cosine')

array([[ 0.97443419]])

Indeed Obama and Clinton are closer (less distance) to one another than Obama and Beckham

#Build a nearest neighbor model for document retrieval

In [35]:
dtm.shape

(59071, 548116)

In [36]:
from sklearn.neighbors import NearestNeighbors
neighbors = NearestNeighbors(n_neighbors=10)

In [37]:
neighbors.fit(dtm)

NearestNeighbors(algorithm='auto', leaf_size=30, metric='minkowski',
         metric_params=None, n_neighbors=10, p=2, radius=1.0)

###Who is closest to Obama?

In [38]:
distances, indices = neighbors.kneighbors(dtm[obama.index], return_distance=True)

[This function](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors.kneighbors) returns an array representing the lengths to the points and an array of indices of the nearest points

In [39]:
distances

array([[ 0.        ,  1.16514466,  1.2073695 ,  1.21963978,  1.22250901,
         1.23617841,  1.24305659,  1.24466747,  1.24829554,  1.25160692]])

In [40]:
indices

array([[35817, 24478, 38376, 57108, 38714, 46140, 18827, 44681,  6796,
         2412]], dtype=int64)

In [41]:
zip(people.iloc[indices[0]]['name'], distances[0])

[('Barack Obama', 0.0),
 ('Joe Biden', 1.16514466168904),
 ('Samantha Power', 1.207369497235288),
 ('Hillary Rodham Clinton', 1.2196397836767043),
 ('Eric Stern (politician)', 1.2225090075525114),
 ('Robert Gibbs', 1.2361784064861681),
 ('Henry Waxman', 1.2430565883007183),
 ('Jesse Lee (politician)', 1.2446674713142944),
 ('Eric Holder', 1.2482955356635752),
 ('Joe the Plumber', 1.2516069176926297)]

Not quite the same results as GraphLab Create, but our #2 is the same, and we got a Clinton (just not the same one). Note that our NearestNeighbors using a minkowski metric instead of a cosine distance metric, so that could be one reason for the difference, among others.

The results do make sense. [Joe the Plumber](https://en.wikipedia.org/wiki/Joe_the_Plumber) actually has quite a bit of references to Obama.

##Other examples of document retrieval

In [42]:
swift = people[people['name'] == 'Taylor Swift']

In [43]:
distances, indices = neighbors.kneighbors(dtm[swift.index], return_distance=True)

In [44]:
zip(people.iloc[indices[0]]['name'], distances[0])

[('Taylor Swift', 1.4901161193847656e-08),
 ('Carrie Underwood', 1.1830037741879376),
 ('Al Swift', 1.1877544656972066),
 ('Ed Sheeran', 1.1939382507224359),
 ('Tim McGraw', 1.197284910223761),
 ('Kelly Clarkson', 1.199188988276866),
 ('Adele', 1.1997900487053563),
 ('Bill Swift', 1.2049650308751825),
 ('Dolly Parton', 1.207080766341732),
 ('Joss Stone', 1.2081386952786064)]

In [45]:
jolie = people[people['name'] == 'Angelina Jolie']

In [46]:
distances, indices = neighbors.kneighbors(dtm[jolie.index], return_distance=True)

In [47]:
zip(people.iloc[indices[0]]['name'], distances[0])

[('Angelina Jolie', 0.0),
 ('Brad Pitt', 1.1739732512825942),
 ('Keith Jolie', 1.241877705023315),
 ('Jodie Foster', 1.2531901283345284),
 ('Maggie Smith', 1.2545732236609843),
 ('Jessica Chastain', 1.2593115814365714),
 ('Anne Hathaway', 1.2601600271995759),
 ('Nicole Kidman', 1.2621055803978345),
 ('Barry Voight', 1.2621427435834212),
 ('Billy Bob Thornton', 1.2638984074312178)]

In [48]:
arnold = people[people['name'] == 'Arnold Schwarzenegger']

In [49]:
distances, indices = neighbors.kneighbors(dtm[arnold.index], return_distance=True)

In [50]:
zip(people.iloc[indices[0]]['name'], distances[0])

[('Arnold Schwarzenegger', 1.4901161193847656e-08),
 ('Bonnie Garcia', 1.2596831766120811),
 ('Paul Grant (bodybuilder)', 1.2632328763204097),
 ('Gray Davis', 1.2838462329808831),
 ('James Tramel', 1.2844632589093408),
 ('Abel Maldonado', 1.28509972154324),
 ('Bruce McPherson', 1.293240247891688),
 ('Charlene Zettel', 1.2941070974369826),
 ('Russell Gould', 1.3016211830154023),
 ('David Israel', 1.3018279325569584)]

The results aren't exactly the same as from GraphLab Create, but they are close. At least they seem to make sense.

If anyone spots any errors or better ways of performing the analysis, feedback is always welcome!

Cheers!

-Stephen